virtcontainers/api.go

```diff
@@ -639,3 +639,25 @@ func ProcessListContainer(sandboxID, containerID string, options ProcessListOpti
 
 	return c.processList(options)
 }
+
+// ContainerSandboxList returns the list of sandboxes IDs for all
+// the sandboxes including the provided container ID.
+// It also provides the information about the container ID being part
+// of the containers map.
+func ContainerSandboxList(containerID string) ([]string, bool, error) {
```

Comment:
TBH, I am still not a fan of this API and the global container-sandbox mapping file. It introduces a global on disk lock and extra json marshaling/unmarshaling for every container creation/deletion. And this extra json marshaling/unmarshaling gets slower as the number of containers increases on the same node. Of course, the extra stateless flag can help frakti/hyperd avoid the situation. But kata cli will still be affected.

I think we should draw wider attention to the change or discuss it in the arch meeting. /cc @sameo @WeiZhang555 @laijs @gnawux @jodh-intel @amshinde @devimc  WDYT? Another option is to implement it in the client side, which does not require global lock nor global mapping file, but it has a drawback that cli need to maintain some container-related information.

Reply:
Backed by @grahamwhaley test results, this change introduces more improvements than it introduces latency.
Also, implementing this at the CLI level does not mean we don't need any locking, the CLI could be called concurrently in order to start several containers on different pods. And in this case you still need a lock to make sure two different `kata-runtime` processes are not trying to write to this file at the same time.

Reply:
Surely this is much faster than current implementation... I was saying it is slower and more complex than being implemented in the cli. And its performance decreases linearly with the number of containers on the host.

> implementing this at the CLI level does not mean we don't need any locking

The containerID uniqueness can be ensured by virtcontainers. Then kata cli only need to record the containerID->sandboxID mapping, which can be done on a per container basis (e.g., using a file like `runtime-root/containerID/sandboxID`). So there is no global mapping file, thus no global disk file locking either.

Reply:
Any implementation will obviously be slower as the number of containers increases.

> The containerID uniqueness can be ensured by virtcontainers

I don't want this since I don't expect this enforcement from virtcontainers. I still want virtcontainers to be able to run 3 different pods with each having some containers with the same IDs. I would expect that Frakti could leverage that for instance. The uniqueness of containerID is really an OCI/CLI requirement only.

> Then kata cli only need to record the containerID->sandboxID mapping, which can be done on a per container basis (e.g., using a file like runtime-root/containerID/sandboxID). So there is no global mapping file, thus no global disk file locking either.

So you're proposing that we're going to create a new file for each new container, which puts more management complexity into the CLI, and that's what I don't really like. But maybe other can give some input here ?

Reply:
> Any implementation will obviously be slower as the number of containers increases.

If you do not maintain a global mapping file, you do not suffer from this.

>  I still want virtcontainers to be able to run 3 different pods with each having some containers with the same IDs.

I expect the same from virtcontainers, -- sandboxID is globally unique, and containerID is unique inside one sandbox.

However for kata cli, we can differentiated docker containers from CRI containers with container annotations.
- For docker containers, containerID uniqueness == sandboxID uniqueness because we always use containerID as sandboxID for docker containers.
- For CRI containers, the uniqueness is ensured by virtcontainers.

So in summary we do not have to check containerID in CLI.

Reply:
@grahamwhaley @jodh-intel @egernst @amshinde WDYT ? The proposal from @bergwolf would work but I have a concern about letting the CLI handling some file creation/removal, this would kind of duplicate the file management between the CLI and virtcontainers.

Reply:
@bergwolf @sboeuf What about supporting several entries per `runtime-root/containerID/` from virtcontainers instead? If there are several of them, then we fallback to the current implementation. Multiple `sanboxID` per `runtime-root/containerID/` should be a very rare occurrence. And since we can safely assume that there will not be concurrent write access to any file under `runtime-root/containerID/` there would not be any need for file locking. Concurrency should only happen at the `runtime-root/containerID/sandboxID` read level.

Reply:
No need to fallback, implementing the same solution inside virtcontainers would work very well. There is only one case that is bothering me, how can we delete `runtime-root/containerID` directory safely ? Because at the time we'll do the delete, we never know that another pod could create the same containerID.

Reply:
I think you can handle this. If you delete a delete a sandbox directory and the check the upper {cid} directory is empty at that point, simply delete it. If another pod has created a directory between the time you delete and check, you are good, you do not delete {cid}.